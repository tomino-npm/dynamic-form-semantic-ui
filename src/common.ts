import { FormElement, DataSet } from '@tomino/dynamic-form';
import { initCss } from '@tomino/toolbelt';

export type SignatureType = {
  comment: string;
  signature: string;
  verifiedState: 'Pending' | 'Verified' | 'Rejected';
  rejected?: boolean;
  uid: string;
  name: string;
  date: Date;
  setValue?(name: string, value: any): any;
};

export type SignatureHandlers = {
  validateForm(): boolean;
  verifySignature(uid: string, signature: string): Promise<string>;
  sign(
    source: string,
    reject: boolean,
    password: string,
    reason: string,
    submit: boolean
  ): Promise<SignatureType>;
  signatureFont(): string;
};

export type FormControlProps = {
  formControl: FormElement;
  owner: DataSet;
  handlers?: { [index: string]: any };
  readOnly: boolean;
};

export const { css, renderCss } = initCss('form-');
