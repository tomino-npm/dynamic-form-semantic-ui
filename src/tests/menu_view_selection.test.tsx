import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { create } from './form_query_data';
import { JSONSchema, FormDefinition, FormModel } from '@tomino/dynamic-form';
import { TestComponent } from './common';
import { clone } from '@tomino/toolbelt';

describe('Menu', () => {
  describe('Selection', () => {
    const formData = {
      description: 'Some Description',
      people: [
        {
          name: 'Tomas Trescak',
          address: 'Sydney, Slovakia'
        },
        {
          name: 'Valeria Toscani',
          address: 'Sydney, Italy'
        },
        {
          name: 'Vittoria Trescakova',
          address: 'Sydney, Australia',
          title: 'Princess'
        }
      ]
    };

    const schema: JSONSchema = {
      type: 'object',
      properties: {
        people: {
          type: 'array',
          items: {
            type: 'object',
            properties: {
              name: { type: 'string' },
              address: { type: 'string' },
              title: { type: 'string' }
            }
          }
        }
      },
      required: ['name']
    };

    const formDefinition: FormDefinition = create.form({
      elements: [
        create.formElement({
          row: 0,
          column: 0,
          width: 16,
          control: 'Menu',
          source: 'people',
          list: 'name',
          controlProps: {
            menuProps: {
              attached: 'top',
              tabular: true
            },
            contentProps: {
              attached: 'bottom'
            }
          },
          elements: [
            {
              control: 'Form',
              elements: [
                {
                  row: 0,
                  column: 0,
                  width: 8,
                  source: 'name',
                  label: 'Name',
                  control: 'Input'
                },
                {
                  row: 0,
                  column: 8,
                  width: 8,
                  source: 'address',
                  label: 'Address',
                  control: 'Input'
                }
              ]
            }
          ]
        })
      ]
    });

    describe('Tabular', () => {
      function component() {
        const form = new FormModel(formDefinition, schema, formData);

        // just another notation
        return <TestComponent form={form} />;
      }

      it('renders tab menu', () => {
        const wrapper = renderer.create(component());
        expect(wrapper).toMatchSnapshot();
      });

      return { component };
    });

    describe('Side Menu', () => {
      function component() {
        let modified: FormDefinition = clone(formDefinition);
        let menu = modified.elements[0];
        menu.controlProps = {
          menuProps: {
            vertical: true,
            secondary: true
          },
          contentProps: {
            style: { float: 'left' }
          }
        };

        const form = new FormModel(modified, schema, formData);

        // just another notation
        return <TestComponent form={form} />;
      }

      it('renders tab menu', () => {
        const wrapper = renderer.create(component());
        expect(wrapper).toMatchSnapshot();
      });

      return { component };
    });

    describe('Top Menu', () => {
      function component() {
        let modified: FormDefinition = clone(formDefinition);
        let menu = modified.elements[0];
        menu.controlProps = {
          menuProps: {
            secondary: true
          }
        };

        const form = new FormModel(modified, schema, formData);

        // just another notation
        return <TestComponent form={form} />;
      }

      it('renders tab menu', () => {
        const wrapper = renderer.create(component());
        expect(wrapper).toMatchSnapshot();
      });

      return { component };
    });
  });
});
