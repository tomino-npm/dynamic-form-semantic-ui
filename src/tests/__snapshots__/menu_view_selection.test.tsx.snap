// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Selection Side Menu renders tab menu 1`] = `
<div
  className="ui segment ui form"
>
  <div
    className="ui dividing header"
  >
    Standard
  </div>
  <div
    className="ui form "
  >
    <div
      className="fields"
    >
      <div
        className="sixteen wide field"
      >
        <div
          className="vertical-bfaadidijh"
        >
          <div
            className="ui secondary vertical menu"
          >
            <a
              className="active item"
              onClick={[Function]}
            >
              Tomas Trescak
            </a>
            <a
              className="item"
              onClick={[Function]}
            >
              Valeria Toscani
            </a>
            <a
              className="item"
              onClick={[Function]}
            >
              Princess
            </a>
          </div>
          <div />
          <div
            className="ui segment"
            style={
              Object {
                "float": "left",
              }
            }
          >
            <div
              className="ui form "
            >
              <div
                className="fields"
              >
                <div
                  className="eight wide field"
                >
                  <label
                    htmlFor="name"
                  >
                    Name
                  </label>
                  <div
                    className="ui input"
                  >
                    <input
                      name="name"
                      onChange={[Function]}
                      type="text"
                      value="Tomas Trescak"
                    />
                  </div>
                </div>
                <div
                  className="eight wide field"
                >
                  <label
                    htmlFor="address"
                  >
                    Address
                  </label>
                  <div
                    className="ui input"
                  >
                    <input
                      name="address"
                      onChange={[Function]}
                      type="text"
                      value="Sydney, Slovakia"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="ui dividing header"
  >
    Readonly
  </div>
  <div
    className="ui form form-hjcaghiia"
  >
    <div
      className="fields"
    >
      <div
        className="sixteen wide field"
      >
        <div
          className="vertical-bfaadidijh"
        >
          <div
            className="ui secondary vertical menu"
          >
            <a
              className="active item"
              onClick={[Function]}
            >
              Tomas Trescak
            </a>
            <a
              className="item"
              onClick={[Function]}
            >
              Valeria Toscani
            </a>
            <a
              className="item"
              onClick={[Function]}
            >
              Princess
            </a>
          </div>
          <div />
          <div
            className="ui segment"
            style={
              Object {
                "float": "left",
              }
            }
          >
            <div
              className="ui form form-hjcaghiia"
            >
              <div
                className="fields"
              >
                <div
                  className="eight wide field"
                >
                  <label
                    htmlFor="name"
                  >
                    Name
                  </label>
                  <div
                    className="ui disabled input"
                  >
                    <input
                      disabled={true}
                      name="name"
                      onChange={[Function]}
                      tabIndex={-1}
                      type="text"
                      value="Tomas Trescak"
                    />
                  </div>
                </div>
                <div
                  className="eight wide field"
                >
                  <label
                    htmlFor="address"
                  >
                    Address
                  </label>
                  <div
                    className="ui disabled input"
                  >
                    <input
                      disabled={true}
                      name="address"
                      onChange={[Function]}
                      tabIndex={-1}
                      type="text"
                      value="Sydney, Slovakia"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Menu Selection Tabular renders tab menu 1`] = `
<div
  className="ui segment ui form"
>
  <div
    className="ui dividing header"
  >
    Standard
  </div>
  <div
    className="ui form "
  >
    <div
      className="fields"
    >
      <div
        className="sixteen wide field"
      >
        <div
          className="none"
        >
          <div
            className="ui top attached tabular menu"
          >
            <a
              className="active item"
              onClick={[Function]}
            >
              Tomas Trescak
            </a>
            <a
              className="item"
              onClick={[Function]}
            >
              Valeria Toscani
            </a>
            <a
              className="item"
              onClick={[Function]}
            >
              Princess
            </a>
          </div>
          <div
            className="ui bottom attached segment"
          >
            <div
              className="ui form "
            >
              <div
                className="fields"
              >
                <div
                  className="eight wide field"
                >
                  <label
                    htmlFor="name"
                  >
                    Name
                  </label>
                  <div
                    className="ui input"
                  >
                    <input
                      name="name"
                      onChange={[Function]}
                      type="text"
                      value="Tomas Trescak"
                    />
                  </div>
                </div>
                <div
                  className="eight wide field"
                >
                  <label
                    htmlFor="address"
                  >
                    Address
                  </label>
                  <div
                    className="ui input"
                  >
                    <input
                      name="address"
                      onChange={[Function]}
                      type="text"
                      value="Sydney, Slovakia"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="ui dividing header"
  >
    Readonly
  </div>
  <div
    className="ui form form-hjcaghiia"
  >
    <div
      className="fields"
    >
      <div
        className="sixteen wide field"
      >
        <div
          className="none"
        >
          <div
            className="ui top attached tabular menu"
          >
            <a
              className="active item"
              onClick={[Function]}
            >
              Tomas Trescak
            </a>
            <a
              className="item"
              onClick={[Function]}
            >
              Valeria Toscani
            </a>
            <a
              className="item"
              onClick={[Function]}
            >
              Princess
            </a>
          </div>
          <div
            className="ui bottom attached segment"
          >
            <div
              className="ui form form-hjcaghiia"
            >
              <div
                className="fields"
              >
                <div
                  className="eight wide field"
                >
                  <label
                    htmlFor="name"
                  >
                    Name
                  </label>
                  <div
                    className="ui disabled input"
                  >
                    <input
                      disabled={true}
                      name="name"
                      onChange={[Function]}
                      tabIndex={-1}
                      type="text"
                      value="Tomas Trescak"
                    />
                  </div>
                </div>
                <div
                  className="eight wide field"
                >
                  <label
                    htmlFor="address"
                  >
                    Address
                  </label>
                  <div
                    className="ui disabled input"
                  >
                    <input
                      disabled={true}
                      name="address"
                      onChange={[Function]}
                      tabIndex={-1}
                      type="text"
                      value="Sydney, Slovakia"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Menu Selection Top Menu renders tab menu 1`] = `
<div
  className="ui segment ui form"
>
  <div
    className="ui dividing header"
  >
    Standard
  </div>
  <div
    className="ui form "
  >
    <div
      className="fields"
    >
      <div
        className="sixteen wide field"
      >
        <div
          className="none"
        >
          <div
            className="ui secondary menu"
          >
            <a
              className="active item"
              onClick={[Function]}
            >
              Tomas Trescak
            </a>
            <a
              className="item"
              onClick={[Function]}
            >
              Valeria Toscani
            </a>
            <a
              className="item"
              onClick={[Function]}
            >
              Princess
            </a>
          </div>
          <div
            className="ui segment"
          >
            <div
              className="ui form "
            >
              <div
                className="fields"
              >
                <div
                  className="eight wide field"
                >
                  <label
                    htmlFor="name"
                  >
                    Name
                  </label>
                  <div
                    className="ui input"
                  >
                    <input
                      name="name"
                      onChange={[Function]}
                      type="text"
                      value="Tomas Trescak"
                    />
                  </div>
                </div>
                <div
                  className="eight wide field"
                >
                  <label
                    htmlFor="address"
                  >
                    Address
                  </label>
                  <div
                    className="ui input"
                  >
                    <input
                      name="address"
                      onChange={[Function]}
                      type="text"
                      value="Sydney, Slovakia"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="ui dividing header"
  >
    Readonly
  </div>
  <div
    className="ui form form-hjcaghiia"
  >
    <div
      className="fields"
    >
      <div
        className="sixteen wide field"
      >
        <div
          className="none"
        >
          <div
            className="ui secondary menu"
          >
            <a
              className="active item"
              onClick={[Function]}
            >
              Tomas Trescak
            </a>
            <a
              className="item"
              onClick={[Function]}
            >
              Valeria Toscani
            </a>
            <a
              className="item"
              onClick={[Function]}
            >
              Princess
            </a>
          </div>
          <div
            className="ui segment"
          >
            <div
              className="ui form form-hjcaghiia"
            >
              <div
                className="fields"
              >
                <div
                  className="eight wide field"
                >
                  <label
                    htmlFor="name"
                  >
                    Name
                  </label>
                  <div
                    className="ui disabled input"
                  >
                    <input
                      disabled={true}
                      name="name"
                      onChange={[Function]}
                      tabIndex={-1}
                      type="text"
                      value="Tomas Trescak"
                    />
                  </div>
                </div>
                <div
                  className="eight wide field"
                >
                  <label
                    htmlFor="address"
                  >
                    Address
                  </label>
                  <div
                    className="ui disabled input"
                  >
                    <input
                      disabled={true}
                      name="address"
                      onChange={[Function]}
                      tabIndex={-1}
                      type="text"
                      value="Sydney, Slovakia"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
